test/apps/pulp-nnx-hal/src/nnx_layer.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Luka Macan <luka.macan@unibo.it>
+ *
+ * Copyright 2023 ETH Zurich and University of Bologna
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "nnx_layer.h"
+#include <pmsis.h>
+
+#ifdef NNX_NE16
+
+#include "ne16.h"
+#include "ne16_task.h"
+
+typedef ne16_norm_mode_e nnx_norm_mode_e;
+typedef ne16_quant_t nnx_quant_t;
+typedef ne16_quant_function_e nnx_quant_function_e;
+typedef ne16_norm_t nnx_norm_t;
+typedef ne16_task_t nnx_task_t;
+typedef ne16_task_flag_e nnx_task_flag_e;
+typedef ne16_task_data_t nnx_task_data_t;
+
+// Minimal copy from ne16/bsp/ne16_pulp_bsp.c
+#define NE16_PULP_BASE_ADDR (0x00201000)
+static const ne16_dev_t nnx_dev = {
+    .hwpe_dev = (struct hwpe_dev_t){
+        .base_addr = (volatile uint32_t *)NE16_PULP_BASE_ADDR}};
+
+#define nnxTaskFlagTrue ne16TaskFlagTrue
+#define nnxTaskFlagFalse ne16TaskFlagFalse
+
+#define nnx_task_init ne16_task_init
+#define nnx_task_set_op_to_conv ne16_task_set_op_to_conv
+#define nnx_task_set_bits ne16_task_set_bits
+#define nnx_task_set_norm_quant ne16_task_set_norm_quant
+#define nnx_task_set_weight_offset ne16_task_set_weight_offset
+#define nnx_task_set_dims ne16_task_set_dims
+#define nnx_task_set_dims_stride2x2 ne16_task_set_dims_stride2x2
+#define nnx_task_set_addr_conv ne16_task_set_addr_conv
+#define nnx_task_set_addr_norm_quant ne16_task_set_addr_norm_quant
+
+#define nnx_task_queue_empty ne16_task_queue_empty
+
+#elif defined NNX_NEUREKA
+
+#include "neureka.h"
+#include "neureka_task.h"
+
+typedef neureka_norm_mode_e nnx_norm_mode_e;
+typedef neureka_quant_t nnx_quant_t;
+typedef neureka_quant_function_e nnx_quant_function_e;
+typedef neureka_norm_t nnx_norm_t;
+typedef neureka_task_t nnx_task_t;
+typedef neureka_task_flag_e nnx_task_flag_e;
+typedef neureka_task_data_t nnx_task_data_t;
+
+// Minimal copy from neureka/bsp/neureka_siracusa_bsp.c
+#define NEUREKA_SIRACUSA_BASE_ADDR (0x00201000)
+static const neureka_dev_t nnx_dev = {
+    .hwpe_dev = (struct hwpe_dev_t){
+        .base_addr = (volatile uint32_t *)NEUREKA_SIRACUSA_BASE_ADDR}};
```

Comment:
Is this base address generic or implementation specific? If it is implementation-specific, you might want to consider wrapping it, so it cna be overwritten without editing a source file.

Reply:
Do you mean just wrapped with `#ifndef`?

Reply:
Wrapped them with `#ifndef`. Commit 06e66bf874b00b233a8d68e2a73d785a509e7b8f